hoomd/md/nlist.py

```diff
@@ -174,251 +158,139 @@ def __init__(self, buffer=0.4, exclusions=('bond',), rebuild_check_delay=1,
-            cls = _hoomd.CellListStencil(hoomd.context.current.system_definition, self.cpp_cl)
-            hoomd.context.current.system.addCompute(cls, self.name + "_cls")
-            self.cpp_nlist = _md.NeighborListGPUStencil(hoomd.context.current.system_definition, 0.0, r_buff, self.cpp_cl, cls)
+    def __init__(self, cell_width, buffer=0.4, exclusions=('bond',),
+                 rebuild_check_delay=1, diameter_shift=False, check_dist=True,
+                 max_diameter=1.0, deterministic=False):
 
-        self.cpp_nlist.setEvery(check_period, dist_check)
-
-        hoomd.context.current.system.addCompute(self.cpp_nlist, self.name)
-        self.cpp_cl.setSortCellList(deterministic)
-
-        # register this neighbor list with the context
-        hoomd.context.current.neighbor_lists += [self]
-
-        # save the user defined parameters
-        self.set_params(r_buff, check_period, d_max, dist_check)
-        self.set_cell_width(cell_width)
-
-    def set_cell_width(self, cell_width):
-        R""" Set the cell width
-
-        Args:
-            cell_width (float): New cell width.
-        """
-        if cell_width is not None:
-            self.cpp_nlist.setCellWidth(float(cell_width))
+        super().__init__(buffer, exclusions, rebuild_check_delay,
+                         diameter_shift, check_dist, max_diameter)
 
-    def tune_cell_width(self, warmup=200000, min_width=None, max_width=None, jumps=20, steps=5000):
-        R""" Make a series of short runs to determine the fastest performing bin width.
+        params = ParameterDict(deterministic=bool(deterministic),
+                               cell_width=float(cell_width))
 
-        Args:
-            warmup (int): Number of time steps to run() to warm up the benchmark
-            min_width (float): Minimum cell bin width to try
-            max_width (float): Maximum cell bin width to try
-            jumps (int): Number of different bin width to test
-            steps (int): Number of time steps to run() at each point
+        self._param_dict.update(params)
 
-        :py:class:`tune_cell_width()` executes *warmup* time steps. Then it sets the nlist *cell_width* value to *min_width* and
-        runs for *steps* time steps. The TPS value is recorded, and the benchmark moves on to the next *cell_width*
-        value completing at *max_width* in *jumps* jumps. Status information is printed out to the screen, and the
-        optimal *cell_width* value is left set for further runs() to continue at optimal settings.
+    def _attach(self):
+        if isinstance(self._simulation.device, hoomd.device.CPU):
+            nlist_cls = _md.NeighborListStencil
+        else:
+            nlist_cls = _md.NeighborListGPUStencil
+        # TODO remove 0.0 r_cut from constructor
+        self._cpp_obj = nlist_cls(self._simulation.state._cpp_sys_def, 0.0,
+                                  self.buffer)
```

Comment:
I can add and export a cpp constructor which would only take 2 arguments so we don't have to always pass this 0 through here. Does this sound like a good idea to others?

Reply:
This would be a bit of a weird pattern because you would be eliminating the middle variable from this full constructor. I prefer explicitly setting r_cut to zero.

Reply:
In v3 API the middle variable (`r_cut`) is more or less a dummy because of how the neighborlist consumer communicates its rcut matrix with the neighborlist. Because of this, there aren't many cases (only cases I'm aware of are cpp unit tests) where this argument is useful on construction of the neighborlists. Because of the above, I don't think this is a weird pattern, its just more in line with how neighborlists are going to be used going forward.

Reply:
I'm pretty sure that variable has been used as a dummy for a while, it is vestigial from when HOOMD did not support per-type cutoffs in the neighbor list. If you don't want to type the zero explicitly, I would propose removing it (and having the class default initialize to 0.0 internally) rather than having two constructors:
```c++
NeighborList(system, r_buff)
```
I find the pattern strange compared to python and most C++ classes where you would want to overload a method with incrementally longer call signatures, but you wouldn't usually want to reorder the variables. This is particularly important here since `r_cut` and `r_buff` have the same type. For example, I would expect:
```c++
NeighborList(system, r_buff);
NeighborList(system, r_buff, r_cut);
```
or even better
```c++
NeighborList(system, r_buff, r_cut=0.0);
```
if you wanted to support both. As you say, there isn't much of a use case for setting `r_cut` though, so the simplest solution would be to remove it.

Reply:
I see what you meant by "weird pattern" originally now. I agree with your last suggestion to reorder the arguments and make the `r_cut` argument default to 0. Removing it would require a lot of work to update the cpp tests, as there are many of them that make use of the single `r_cut` and don't have consumers that communicate matrices with the neighborlists.

Reply:
But, if you reorder the arguments, won't you have to update the unit tests?

I think it would be find to remove rcut entirely. You can rewrite the unit tests with a global regex search and replace.